cmake/modules/extensions.cmake

```diff
@@ -1468,19 +1468,34 @@ endfunction()
 #
 # This is a common function to ensure that build strings are always created
 # in a uniform way.
+# A single string is returned containing the full build string constructed from
+# all arguments.
+#
+# When MERGE is supplied a list of build strings will be returned with the full
+# build string as first item in the list.
+# The full order of build strings returned in the list will be:
+# - Full build string, including identifier revision, build type
+# - Build string with build type removed
+# - Build string with revision removed in addition
+# - Build string with board variants removed in addition
+# - Build string with cpuset removed in addition
+# - Build string with soc removed in addition
```

Comment:
Based on this description, I tried:
```cmake
zephyr_build_string(build_string BOARD alpha BOARD_REVISION 1.0.0 BOARD_IDENTIFIER /soc/bar BUILD foo MERGE)

# I expected:  alpha_soc_bar_1_0_0_foo;alpha_soc_bar_1_0_0;alpha_soc_bar;alpha_soc;alpha
# but I got:   alpha_soc_bar_1_0_0_foo;alpha_soc_bar;alpha
```

It works if I streamline the implementation like this:
```cmake
  string(REPLACE "/" ";" build_str_segments "${BUILD_STR_BOARD}${BUILD_STR_BOARD_IDENTIFIER}")
  if(DEFINED BUILD_STR_BOARD_REVISION)
    string(REPLACE "." "_" revision_string ${BUILD_STR_BOARD_REVISION})
    list(APPEND build_str_segments ${revision_string})
  endif()
  list(APPEND build_str_segments ${BUILD_STR_BUILD})

  while(NOT build_str_segments STREQUAL "")
    list(JOIN build_str_segments "_" variant_string)
    list(APPEND ${outvar} "${variant_string}")
    list(POP_BACK build_str_segments)
    if(NOT BUILD_STR_MERGE)
      break()
    endif()
  endwhile()
```

Reply:
Nice catch 👍 
And thanks for the proposal, using `JOIN` sure is cleaner.

Reply:
fixed, resorted to use `string(JOIN ...)` as that allows for `BUILD_STR_BUILD` at end, so that all build strings are appended with that info when passed as argument.

That conforms to what we do today with build string, and it was an oversight in first round that it was only appended to first string.

Function description updated as well.